### PR TITLE
feat(pytorch): Allow TransformerLayer and MultiheadAttention to accept sequence length parameters

### PR DESCRIPTION
**Pull request by hXl3s**
# Description

TransformerLayer and MultiheadAttention does not allow for passing arbitrary length sequences. While this feature is supported by DotProductAttention it cannot be controlled by higher abstraction layers. This PR fixes that issue.

Additionally fixes runtime bug when thd layout is used

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Added parameters `cu_seqlen_q`, `cu_seqlens_kv`, `max_seqlen_q` and `max_seqlen_kv` to MultiheadAttention and TransformerLayer
- When using THD layout for attention input, shapes are handled correctly

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ptrendx** [comment]
Hi @hXl3s please sign your commits. See https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst#sign-your-work for details.

**Comment by ptrendx** [comment]
Could you also add an error when somebody tries to run THD in inference with KV-cache here: https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/attention.py#L6732-L6735?
@sudhakarsingh27 FYI, since this will touch the pieces you are looking at for merging THD inference support.

**Comment by ptrendx** [comment]
Also @hXl3s could you add some unit test to make sure it works (e.g. comparing thd vs bshd outputs of TransformerLayer)?

**Comment by hXl3s** [comment]
@ptrendx Added to testcase comparing output of THD vs BSHD for float16 and bfloat16. 
float32 is skipped as apparently cudnn or other implementation used does not support THD with float32

Additionally, as requested there is an assert now, that check if you do not use THD layout with KV-Cache during inference

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -6459,6 +6459,10 @@ def forward(
         core_attention_bias_type: str = "no_bias",
         core_attention_bias: Optional[torch.Tensor] = None,
         alibi_slopes: Optional[torch.Tensor] = None,
+        cu_seqlens_q: Optional[torch.Tensor] = None,
+        cu_seqlens_kv: Optional[torch.Tensor] = None,
+        max_seqlen_q: int = None,
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:6464]: Those should also be Optional, right?

File: transformer_engine/pytorch/transformer.py [hunk 2]
```diff
@@ -525,6 +525,10 @@ def forward(
         core_attention_bias_type: str = "no_bias",
         core_attention_bias: Optional[torch.Tensor] = None,
         alibi_slopes: Optional[torch.Tensor] = None,
+        cu_seqlens_q: Optional[torch.Tensor] = None,
+        cu_seqlens_kv: Optional[torch.Tensor] = None,
+        max_seqlen_q: int = None,
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:530]: Same

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -6639,6 +6658,11 @@ def forward(
                 for x in (query_layer, key_layer, value_layer)
             )
 
+            if self.qkv_format == "thd":
+                query_layer, key_layer, value_layer = (
+                    x.reshape(x.size(0), x.size(2), x.size(3))
+                    for x in (query_layer, key_layer, value_layer)
+                )
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:6665]: Seems strange to first always do the sbhd/bshd logic and then trying overwriting it with thd logic. Could we do
```python
            if self.qkv_format == "thd":
                query_layer, key_layer, value_layer = (
                    x.reshape(x.size(0), x.size(2), x.size(3))
                    for x in (query_layer, key_layer, value_layer)
                )
            else:
                query_layer, key_layer, value_layer = (
                    x.reshape(x.size(0), x.size(1), -1, self.hidden_size_per_attention_head)
                    for x in (query_layer, key_layer, value_layer)
                )
```
instead?

**Comment by hXl3s** [comment]
Resolved all comments

**Review by ptrendx — APPROVED** [review]
LGTM, thanks!

**Comment by ptrendx** [comment]
@hXl3s Could you skip the thd test when GPU SM arch is lower than 8.0 (as neither FlashAttention nor cuDNN support those).

**Pull request merged**